dm/syncer/binlogstream/streamer_controller.go

```diff
@@ -333,35 +354,41 @@ func (c *StreamerController) GetEvent(tctx *tcontext.Context) (*replication.Binl
 
 	if err != nil {
 		if err == context.Canceled || err == context.DeadlineExceeded {
-			return nil, 0, pb.ErrorOp_InvalidErrorOp, err
+			return nil, pb.ErrorOp_InvalidErrorOp, err
 		}
 		c.Lock()
 		tctx.L().Error("meet error when get binlog event", zap.Error(err))
 		c.meetError = true
 		c.Unlock()
-		return nil, 0, pb.ErrorOp_InvalidErrorOp, err
+		return nil, pb.ErrorOp_InvalidErrorOp, err
 	}
 
 	if suffix != 0 {
 		c.locations.curStartLocation = c.upstream.curStartLocation
 		c.locations.curStartLocation.Suffix = suffix - 1
 		c.locations.curEndLocation = c.upstream.curStartLocation
 		c.locations.curEndLocation.Suffix = suffix
+		// we only allow injecting DDL, so txnEndLocation is end location of every injected event
 		c.locations.txnEndLocation = c.locations.curEndLocation
-	} else {
-		if c.locations.curEndLocation.Suffix != 0 {
-			// this is the first upstream event after injection, keep suffix for start location
-			c.locations.curStartLocation = c.locations.curEndLocation
-			c.locations.curEndLocation = c.upstream.curEndLocation
-			c.locations.txnEndLocation = c.upstream.txnEndLocation
-		} else {
-			c.locations = c.upstream.locations
-		}
+		return event, op, nil
 	}
 
-	return event, suffix, op, nil
+	if isDataEvent(event) {
+		c.locations.curStartLocation = c.locations.curEndLocation
+		c.locations.curEndLocation = c.upstream.curEndLocation
+		c.locations.txnEndLocation = c.upstream.txnEndLocation
+		return event, op, nil
+	}
+
+	c.locations.curStartLocation = c.locations.curEndLocation
+	c.locations.curEndLocation.CopyWithoutSuffixFrom(c.upstream.curEndLocation)
```

Comment:
when will loc of upstream has suffix?

Reply:
upstream locations always has suffix 0, only injected events has non-zero suffix. When we reset to the middle of injected events whose suffix is not 0, upstream locations will overwrite the suffix without `CopyWithoutSuffixFrom`